src/python/pants/engine/fs_test.py

```diff
@@ -142,6 +142,18 @@ def try_with_backoff(assertion_fn: Callable[[], bool], count: int = 4) -> bool:
     return False
 
 
+# -----------------------------------------------------------------------------------------------
+# `FileContent`
+# -----------------------------------------------------------------------------------------------
+
+
+def test_file_content_non_bytes():
```

Comment:
If you add types to this test function signature, you'll see why `FileContent.content` does not need runtime type checking. Pants leans on build time type checking, and if you're writing plugin code, so should you:
```
$ pants check ::
03:44:03.48 [INFO] Completed: Check compilation for Scala - scalac - scalac succeeded.
03:44:03.48 [INFO] Completed: Check javac compilation - javac - javac succeeded.
03:44:03.52 [INFO] Completed: Typecheck using MyPy - mypy - mypy succeeded.
Success: no issues found in 1578 source files

✓ javac succeeded.
✓ mypy succeeded.
✓ scalac succeeded.
$ pants test src/python/pants/engine/fs_test.py -- -k test_file_content_non_bytes
03:44:06.41 [INFO] Completed: Run Pytest - src/python/pants/engine/fs_test.py:tests - succeeded.

✓ src/python/pants/engine/fs_test.py:tests succeeded in 1.02s (memoized).
$ vi src/python/pants/engine/fs_test.py
jsirois@Gill-Windows:~/dev/pantsbuild/pants (main *) $ git diff
diff --git a/src/python/pants/engine/fs_test.py b/src/python/pants/engine/fs_test.py
index b4d5efa667..7e8099b38e 100644
--- a/src/python/pants/engine/fs_test.py
+++ b/src/python/pants/engine/fs_test.py
@@ -147,7 +147,7 @@ def try_with_backoff(assertion_fn: Callable[[], bool], count: int = 4) -> bool:
 # -----------------------------------------------------------------------------------------------


-def test_file_content_non_bytes():
+def test_file_content_non_bytes() -> None:
     with pytest.raises(TypeError) as exc:
         FileContent(path="4.txt", content="four")

jsirois@Gill-Windows:~/dev/pantsbuild/pants (main *) $ pants check ::
03:44:39.23 [INFO] Completed: Check compilation for Scala - scalac - scalac succeeded.
03:44:39.24 [INFO] Completed: Check javac compilation - javac - javac succeeded.
03:44:42.35 [ERROR] Completed: Typecheck using MyPy - mypy - mypy failed (exit code 1).
src/python/pants/engine/fs_test.py: note: In function "test_file_content_non_bytes":
src/python/pants/engine/fs_test.py:152:43: error: Argument "content" to "FileContent" has incompatible type "str"; expected "bytes"  [arg-type]
            FileContent(path="4.txt", content="four")
                                              ^~~~~~
Found 1 error in 1 file (checked 1578 source files)

✓ javac succeeded.
✕ mypy failed.
✓ scalac succeeded.
```

Reply:
We can document that plugin owners should typecheck their plugins, but if they don't (and I can't assess whether that is an easy thing for them to do, not knowing their codebases) they may run into this issue, get an unintuitive infinite loop and reach out to us for support. This PR helps prevent that. 

I agree that runtime type checking is not great and is to be used sparingly at best, but since this specific case has cropped up more than once, and imposes a support burden on maintainers, it seems like the right tradeoff to me.

Plugin development is already difficult for the casual user, so if we can remove one landmine it seems worth doing.

Reply:
What I'm hearing is post-hoc rationalization. But maybe I'm hearing wrong. Surely the production code added which lacks typing (` -> None`) should be typed and the test function commented to explain why it's not typed. As to people being burdened by type checking plugin code and Pants in this way; well I've left the project over continuous muddled thinking. My reading of the contribution is that it is from a user not accustomed to type checking; after all it's easy to do in this context since it's part of Pants CI and yet still missed 2x in not many more lines of code than that.

Reply:
Interesting - adding the output type of `None` as shown above appears to be a weird quirk with `mypy` in terms of getting the type checker to correctly yell. Using `pyright` (which is the type checker I use for my own projects), I get the same error even without adding that output type.

That feels particularly brittle in terms of how the issue would get identified at build time. Most of my own code is build-time checked but it doesn't inspire confidence that different type checkers do different things with the same code.